File: apps/viewer/src/graph/trace-status.ts
import type { RuntimeEvent } from "../types";

const problemStatuses = ["failed", "timeout", "cancelled", "policy_blocked", "degraded"];

export function traceHasProblem(event: RuntimeEvent): boolean {
  return problemStatuses.includes(event.status ?? "");
}

export function traceStatus(events: RuntimeEvent[]): string {
  const problem = events.find(traceHasProblem);
  return problem?.status ?? "ok";
}

export function traceDuration(events: RuntimeEvent[]): number {
  return events.reduce((sum, event) => sum + (event.duration_ms ?? 0), 0);
}

File: crates/dslraid-cli/src/commands/coverage/overlay/trace_status.rs
use serde_json::Value;

pub(super) fn event_failed(event: &Value, kind: &str) -> bool {
    event
        .get("status")
        .and_then(Value::as_str)
        .is_some_and(|status| {
            matches!(
                status,
                "failed" | "timeout" | "cancelled" | "policy_blocked" | "degraded"
            )
        })
        || kind == "transition_failed"
}
